Documentation and minor plugin changes. (#1573)
freechipsproject/chisel3: docs/src/explanations/naming.md

---
layout: docs
title:  "Naming"
section: "chisel3"
---

Historically, Chisel has had trouble reliably capturing the names of signals. The reasons for this are due to (1)
primarily relying on reflection to find names, (2) using `@chiselName` macro which had unreliable behavior, and (3)

> when all else fails using the `.suggestName("foo")` API.

Reviewer: this isn't exactly a "reason" for why Chisel couldn't reliably capture the names of signals

> With Chisel 3.4 and later, a custom Scala compiler plugin was introduced which enabled reliably and automatically
> capturing the names of signals, when they are declared. In addition, this release included prolific use of a new

Reviewer: ```suggestion
With Chisel 3.4 and later, a custom Scala compiler plugin was introduced which enabled reliable and automatic
capturing of signal names, when they are declared. In addition, this release included prolific use of a new
```

prefixing API which enabled more stable naming of signals programmatically generated from function calls.

This document explains how naming now works in Chisel for signal and module names. For cookbook examples on how to fix
systemic name-stability issues, please refer to the naming [cookbook](../cookbooks/naming.md).

### Compiler Plugin

```scala mdoc:invisible
import chisel3.plugin._
import chisel3._
import chisel3.experimental.prefix
import chisel3.experimental.noPrefix

> import chisel3.stage.ChiselStage

Reviewer 2: I'm not sure this should be invisible, or at least, not all of it. We talk about using `prefix` and `noPrefix` below but then hide the imports to use them.

```

> With the release of Chisel 3.4, all users must add the following line to their build.sbt settings to get the improved

Reviewer 2: ```suggestion
With the release of Chisel 3.4, users should add the following line to their build.sbt settings to get the improved
```
Nit, but seemed a little aggressive 😉

line above is rewritten to.

If the line is within a bundle declaration or is a module instantiation, it is rewritten to replace the right hand
side with a call to `autoNameRecursively`, which names the signal/module.

```scala mdoc
class MyBundle extends Bundle {
  val foo = Input(UInt(3.W))
  // val foo = autoNameRecursively("foo", Input(UInt(3.W)))
}
class Example1 extends MultiIOModule {
  val io = IO(new MyBundle())
  // val io = autoNameRecursively("io", IO(new MyBundle()))
}
println(ChiselStage.emitVerilog(new Example1))
```

Otherwise, it is rewritten to also include the name as a prefix to any signals generated while executing the right-hand-
side of the val declaration:

```scala mdoc
class Example2 extends MultiIOModule {
  val in = IO(Input(UInt(2.W)))
  // val in = autoNameRecursively("in", prefix("in")(IO(Input(UInt(2.W)))))

  val out = IO(Output(UInt(2.W)))
  // val out = autoNameRecursively("out", prefix("out")(IO(Output(UInt(2.W)))))

  def inXin() = in * in

  val add = 3.U + inXin()
  // val add = autoNameRecursively("add", prefix("add")(3.U + inXin()))

>   // Note that the intermediate result of the first addition is prefixed with `add`

Reviewer 2: ```suggestion
  // Note that the intermediate result of the multiplication is prefixed with `add`
```

  out := add + 1.U
}

println(ChiselStage.emitVerilog(new Example2))
```

> Note that the naming also works if the hardware type is nested in an `Option` or an `Iterable`:

Reviewer 2: ```suggestion
Note that the naming also works if the hardware type is nested in an `Option` or a subtype of `Iterable`:
```
Might help people less familiar with the Scala collections library. We could also have a reference to the ScalaDoc here but not needed.

```scala mdoc
class Example8 extends MultiIOModule {
  val in = IO(Input(UInt(2.W)))
  val out = IO(Output(UInt()))

  val add = (in + (in + in).suggestName("foo"))

  out := add
}

println(ChiselStage.emitVerilog(new Example8))
```

### Set a Module Name

If you want to specify the module's name (not the instance name of a module), you can always override the `desiredName`
value. Note that you can parameterize the name by the module's parameters. This is an excellent way to make your module
names more stable and is highly recommended to do.

```scala mdoc
class Example9(width: Int) extends MultiIOModule {
  override val desiredName = s"EXAMPLE9WITHWIDTH$width"
  val in = IO(Input(UInt(width.W)))
  val out = IO(Output(UInt()))

  val add = (in + (in + in).suggestName("foo"))

  out := add
}

println(ChiselStage.emitVerilog(new Example9(8)))
println(ChiselStage.emitVerilog(new Example9(1)))
```

### Reflection Naming

Regardless of whether the compiler plugin is enabled or not, after Chisel constructs a module, it attempts to name all
members of the Module. This will name all vals which are fields of the module class, but it will not name any
vals in nested functions or scopes.

> 

Reviewer 2: An example would be helpful to show why it's so limited.
